### ISSUE TITLE
`semi`: `"beforeStatementContinuationChars": "always"` does not always require semicolon

### ISSUE THREAD

**Issue by andjsrk** [eslint-stylistic/eslint-stylistic#668]
### Validations

- [x] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [x] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

```js
/* eslint @stylistic/semi: ["error", "never", { "beforeStatementContinuationChars": "always" }] */

// 1) currently reports missing semicolon (expected)
do {} while (false)
[1, 2, 3].forEach()

// 2) currently reports nothing (unexpected)
function foo() {}
[1, 2, 3].forEach()
// or
foo()
[1, 2, 3].forEach()
// or
const bar = foo()
[1, 2, 3].forEach()
// etc
```
Note that `2)` is the same as the original ESLint behavior (https://github.com/eslint/eslint/issues/11053#issuecomment-435658719), but it *is not documented* (the document simply says "`"beforeStatementContinuationChars": "always"` requires semicolons at the end of statements if the next line starts with `[`, `(`, `/`, `+`, or `-`.").

I'm actually using rule `semi-style` with option `"first"` so the code with semicolons looks like:
```js
do {} while (false)
;[1, 2, 3].forEach()

function foo() {}
;[1, 2, 3].forEach()

foo()
;[1, 2, 3].forEach()
```
So in my case the configuration `"never", { "beforeStatementContinuationChars": "always" }` keeps consistency for these ASI hazard exceptions so let it improve portability(IDK this is an appropriate word though) of the statement.
I want to keep this consistency regardless of the existence of ESLint rule [`no-unexpected-multiline`](https://eslint.org/docs/latest/rules/no-unexpected-multiline).

### Reproduction

https://stackblitz.com/edit/github-tzngqxaz

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [ ] I am willing to submit a PR with failing tests
